### PR TITLE
mysql-shell: init at 8.0.28

### PR DESCRIPTION
**Pull request by aaronjheng**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -575,6 +575,16 @@ with pkgs;
...
   mod = callPackage ../development/tools/mod { };
 
+  mysql-shell = callPackage ../development/tools/mysql-shell {
+    inherit (darwin) cctools developer_cmds DarwinTools;
+    inherit (darwin.apple_sdk.frameworks) CoreServices;
+    boost = boost173; # Configure checks for specific version.
+    protobuf = protobuf3_11;
+    icu =  icu67;
+    v8 = v8_8_x;
+    python = python39;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:585]:
```suggestion
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+  '';
+
+  meta = with lib;
+    {
+      homepage = "https://dev.mysql.com/doc/mysql-shell/8.0/en/";
+      description = "MySQL Shell is a new command line scriptable shell for MySQL. It supports JavaScript and Python.";
+      license = licenses.gpl2;
+      maintainers = with maintainers; [ aaronjheng ];
+      mainProgram = "mysqlsh";
+    };
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:173]:
```suggestion
  meta = with lib; {
    homepage = "https://dev.mysql.com/doc/mysql-shell/8.0/en/";
    description = "MySQL Shell is a new command line scriptable shell for MySQL. It supports JavaScript and Python.";
    license = licenses.gpl2;
    maintainers = with maintainers; [ aaronjheng ];
    mainProgram = "mysqlsh";
  };
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  installPhase = ''
+    runHook preInstall
+
+    cmake --install ../mysql-shell-${version}-build --prefix $out
+
+    echo '${site}' >> $out/lib/mysqlsh/python-packages/mysqlsh/__init__.py
+
+    runHook postInstall
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:164]: Can't we use the default cmake phases?

File: pkgs/development/tools/mysql-shell/default.nix [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      -DWITH_UNIT_TESTS=OFF \
+      -DFORCE_UNSUPPORTED_COMPILER=1 \
+      -S ../mysql-${version} -B ../mysql-${version}-build
+
+    cmake --build ../mysql-${version}-build --parallel ''${NIX_BUILD_CORES:-1} --target mysqlclient mysqlxclient
+
+    export CXXFLAGS="-DV8_COMPRESS_POINTERS=1 -DV8_31BIT_SMIS_ON_64BIT_ARCH=1"
+
+    mkdir -p ../v8/lib
+    cp ${v8}/lib/libv8.a ../v8/lib/libv8_monolith.a
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:133]:
```suggestion
    ln -s ${v8}/lib/libv8.a ../v8/lib/libv8_monolith.a
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+  dontConfigure = true;
+
+  preBuild = ''
+    cmake -DWITH_BOOST=system \
+      -DWITH_SYSTEM_LIBS=ON \
+      -DWITH_ROUTER=OFF \
+      -DWITH_UNIT_TESTS=OFF \
+      -DFORCE_UNSUPPORTED_COMPILER=1 \
+      -S ../mysql-${version} -B ../mysql-${version}-build
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:126]: cmake flags?

File: pkgs/development/tools/mysql-shell/default.nix [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      fetchurl {
+        url = "https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-${version}.tar.gz";
+        sha256 = "sha256-2Gk2nrbeTyuy2407Mbe3OWjjVuX/xDVPS5ZlirHkiyI=";
+      }
+    )
+  ];
+
+  sourceRoot = "mysql-shell-${version}-src";
+
+  patchPhase = ''
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:70]:
```suggestion
  postPatch = ''
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        url = "https://cdn.mysql.com//Downloads/MySQL-Shell/mysql-shell-${version}-src.tar.gz";
+        sha256 = "sha256-xm2sepVgI0MPs25vu+BcRQeksaVhHcQlymreN1myu6c=";
+      }
+    )
+    (
+      fetchurl {
+        url = "https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-${version}.tar.gz";
+        sha256 = "sha256-2Gk2nrbeTyuy2407Mbe3OWjjVuX/xDVPS5ZlirHkiyI=";
+      }
+    )
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:65]:
```suggestion
    (fetchurl {
      url = "https://cdn.mysql.com//Downloads/MySQL-Shell/mysql-shell-${version}-src.tar.gz";
      sha256 = "sha256-xm2sepVgI0MPs25vu+BcRQeksaVhHcQlymreN1myu6c=";
    })
    (fetchurl {
      url = "https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-${version}.tar.gz";
      sha256 = "sha256-2Gk2nrbeTyuy2407Mbe3OWjjVuX/xDVPS5ZlirHkiyI=";
    })
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+}:
+
+with builtins;
+let
+  pythonDeps = [
+    python.pkgs.certifi
+    python.pkgs.paramiko
+  ];
+  site = ''
+
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:45]:
```suggestion
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    import sys; sys.path.extend([${concatStringsSep ", " (map (x: ''"${x}/${python.sitePackages}"'') pythonDeps)}])
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:46]: Why are we not using the standard python tooling?

File: pkgs/development/tools/mysql-shell/default.nix [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+, numactl
+, cctools
+, CoreServices
+, developer_cmds
+, DarwinTools
+, testVersion
+, mysql-shell
+}:
+
+with builtins;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:38]:
```suggestion
```
Please prefix `lib.` where necessary.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (aaronjheng) [reply pkgs/top-level/all-packages.nix:585]: It needs to be Python 3. How about `python = python3` ?

File: pkgs/development/tools/mysql-shell/default.nix (hunk 8 above)
Comment (aaronjheng) [reply pkgs/development/tools/mysql-shell/default.nix:45]: It's on purpose. It appends `site` content to a python file.

File: pkgs/development/tools/mysql-shell/default.nix (hunk 9 above)
Comment (aaronjheng) [reply pkgs/development/tools/mysql-shell/default.nix:46]: Mysql Shell embeds the python runtime. It's not a typical python package. Any suggestions ?

File: pkgs/development/tools/mysql-shell/default.nix (hunk 3 above)
Comment (aaronjheng) [reply pkgs/development/tools/mysql-shell/default.nix:164]: Before build MySQL Shell, We should build MySQL Server manually. I don't know how to do two cmake builds. So I choose to do this manually. Any suggestions?

File: pkgs/development/tools/mysql-shell/default.nix [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  pname = "mysql-shell";
+  version = "8.0.28";
+
+  srcs = [
+    (fetchurl {
+      url = "https://cdn.mysql.com//Downloads/MySQL-Shell/mysql-shell-${version}-src.tar.gz";
+      sha256 = "sha256-xm2sepVgI0MPs25vu+BcRQeksaVhHcQlymreN1myu6c=";
+    })
+    (fetchurl {
+      url = "https://dev.mysql.com/get/Downloads/MySQL-8.0/mysql-${version}.tar.gz";
```
Comment (jtojnar) [inline pkgs/development/tools/mysql-shell/default.nix:58]: Is not it possible to use `mysqlclient` from `mysql80` package?

Comment (aaronjheng) [reply pkgs/development/tools/mysql-shell/default.nix:58]: No. MySQL Shell needs `mysqlclient` and `mysqlxclient`. There are some tedious rules in MySQL Shell's build scripts which make it hard to use the libraries from `mysql80` package. As far as I can tell.

File: pkgs/development/tools/mysql-shell/default.nix [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:58]:
```suggestion
      url = "https://dev.mysql.com/get/Downloads/MySQL-${lib.versions.majorMinor version}/mysql-${version}.tar.gz";
```

File: pkgs/development/tools/mysql-shell/default.nix [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    })
+  ];
+
+  sourceRoot = "mysql-shell-${version}-src";
+
+  postPatch = ''
+    patch /build/mysql-${version}/cmake/fido2.cmake ${./fido2.cmake.patch}
+
+    substituteInPlace /build/mysql-${version}/cmake/libutils.cmake --replace /usr/bin/libtool libtool
+    substituteInPlace /build/mysql-${version}/cmake/os/Darwin.cmake --replace /usr/bin/libtool libtool
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:69]: We shouldn't reference `/build`. Please use a relative path.

File: pkgs/top-level/all-packages.nix [hunk 14]
```diff
@@ -593,6 +593,16 @@ with pkgs;
...
   mod = callPackage ../development/tools/mod { };
 
+  mysql-shell = callPackage ../development/tools/mysql-shell {
+    inherit (darwin) cctools developer_cmds DarwinTools;
+    inherit (darwin.apple_sdk.frameworks) CoreServices;
+    boost = boost173; # Configure checks for specific version.
+    protobuf = protobuf3_11;
+    icu =  icu67;
+    v8 = v8_8_x;
+    python = python3;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:603]: If the package is not part of pythonPackages we can just use python3 as an input.

File: pkgs/development/tools/mysql-shell/default.nix [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+  srcs = [
+    (fetchurl {
+      url = "https://cdn.mysql.com/get/Downloads/MySQL-Shell/mysql-shell-${version}-src.tar.gz";
+      sha256 = "sha256-xm2sepVgI0MPs25vu+BcRQeksaVhHcQlymreN1myu6c=";
+    })
+    (fetchurl {
+      url = "https://dev.mysql.com/get/Downloads/MySQL-${lib.versions.majorMinor version}/mysql-${version}.tar.gz";
+      sha256 = "sha256-2Gk2nrbeTyuy2407Mbe3OWjjVuX/xDVPS5ZlirHkiyI=";
+    })
```
Comment (SuperSandro2000) [inline pkgs/development/tools/mysql-shell/default.nix:60]:
```
trying https://cdn.mysql.com/get/Downloads/MySQL-Shell/mysql-shell-8.0.28-src.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0    10    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
error: cannot download mysql-shell-8.0.28-src.tar.gz from any mirr
```

**Pull request merged**